src/lib/cache/cache.hpp

```diff
@@ -24,69 +24,90 @@ class Cache {
   explicit Cache(size_t capacity = DefaultCacheCapacity)
       : _impl(std::move(std::make_unique<GDFSCache<Key, Value>>(capacity))) {}
 
-  virtual ~Cache() {}
-
   // Adds or refreshes the cache entry [query, value].
   void set(const Key& query, const Value& value) {
+    std::unique_lock<std::shared_mutex> lock(_mutex);
+
     if (_impl->capacity() == 0) return;
 
-    std::lock_guard<std::mutex> lock(_mutex);
     _impl->set(query, value);
   }
 
-  // Tries to fetch the cache entry for the query into the result object.
-  // Returns true if the entry was found, false otherwise.
+  // Tries to fetch the cache entry for the query into the result object. Returns true if the entry was found, false
+  // otherwise. This needs a write lock to be acquired as most implementation update some type of access count when
+  // retrieving an entry.
   std::optional<Value> try_get(const Key& query) {
+    std::unique_lock<std::shared_mutex> lock(_mutex);
+
     if (_impl->capacity() == 0) return {};
 
-    std::lock_guard<std::mutex> lock(_mutex);
     if (!_impl->has(query)) {
       return {};
     }
     return _impl->get(query);
   }
 
   // Checks whether an entry for the query exists.
-  bool has(const Key& query) const { return _impl->has(query); }
+  bool has(const Key& query) const {
+    std::shared_lock<std::shared_mutex> lock(_mutex);
+    return _impl->has(query);
+  }
 
-  // Returns and refreshes the cache entry for the given query.
-  // Causes undefined behavior if the query is not in the cache.
+  // Returns and refreshes the cache entry for the given query. Causes undefined behavior if the query is not in the
+  // cache. This needs a write lock to be acquired as most implementation update some type of access count when
+  // retrieving an entry.
+  // TODO(anyone): Change this to an assertion
   Value get_entry(const Key& query) {
-    std::lock_guard<std::mutex> lock(_mutex);
+    std::unique_lock<std::shared_mutex> lock(_mutex);
     return _impl->get(query);
   }
 
   // Purges all entries from the cache.
-  void clear() { _impl->clear(); }
+  void clear() {
+    std::unique_lock<std::shared_mutex> lock(_mutex);
+    _impl->clear();
+  }
 
-  void resize(size_t capacity) { _impl->resize(capacity); }
+  void resize(size_t capacity) {
+    std::unique_lock<std::shared_mutex> lock(_mutex);
+    _impl->resize(capacity);
+  }
 
-  // Returns the access frequency of a cached item (=1 for set(), +1 for each get()).
-  // Returns 0 for keys not being cache-resident.
-  size_t size() const { return _impl->size(); }
+  size_t size() const {
+    std::shared_lock<std::shared_mutex> lock(_mutex);
+    return _impl->size();
+  }
 
   // Returns a reference to the underlying cache.
-  AbstractCacheImpl<Key, Value>& cache() { return *_impl; }
-  const AbstractCacheImpl<Key, Value>& cache() const { return *_impl; }
+  // TODO: Is this actually needed?
```

Comment:
I did not follow the whole "caches no longer global" discussion. Is this a place where we need this access?

Reply:
You tell me. The code base doesn't use it. If you and @Bensk1 don't use it in your code, I'd throw it out

Reply:
In my plugin, I use `for (const auto& [query_string, physical_query_plan] : *SQLPipelineBuilder::default_pqp_cache)`. Can't find a use of `cache()`.

@Bensk1 ?